src/shogun/machine/gp/FITCInferenceBase.h

```diff
@@ -0,0 +1,291 @@
+{
+
+/** @brief The Fully Independent Conditional Training inference base class.
+ *
+ * This inference method computes the Cholesky and Alpha vectors approximately
+ * with the help of latent variables. For more details, see "Sparse Gaussian
+ * Process using Pseudo-inputs", Edward Snelson, Zoubin Ghahramani, NIPS 18, MIT
+ * Press, 2005.
+ *
+ * This specific implementation was inspired by the infFITC.m file in the GPML
+ * toolbox.
+ *
+ */
+class CFITCInferenceBase: public CInferenceMethod
+{
+public:
+	/** default constructor */
+	CFITCInferenceBase();
+
+	/** constructor
+	 *
+	 * @param kernel covariance function
+	 * @param features features to use in inference
+	 * @param mean mean function
+	 * @param labels labels of the features
+	 * @param model likelihood model to use
+	 * @param latent_features features to use
+	 */
+	CFITCInferenceBase(CKernel* kernel, CFeatures* features,
+			CMeanFunction* mean, CLabels* labels, CLikelihoodModel* model,
+			CFeatures* latent_features);
+
+	virtual ~CFITCInferenceBase();
+
+	/** return what type of inference we are
+	 *
+	 * @return inference type FITC
+	 */
+	virtual EInferenceType get_inference_type() const { return INF_FITC; }
+
+	/** returns the name of the inference method
+	 *
+	 * @return name FITCBase
+	 */
+	virtual const char* get_name() const { return "FITCBaseInferenceMethod"; }
+
+	/** set latent features
+	 *
+	 * @param feat features to set
+	 */
+	virtual void set_latent_features(CFeatures* feat)
+	{
+		SG_REF(feat);
+		SG_UNREF(m_latent_features);
+		m_latent_features=feat;
+	}
+
+	/** get latent features
+	 *
+	 * @return features
+	 */
+	virtual CFeatures* get_latent_features()
+	{
+		SG_REF(m_latent_features);
+		return m_latent_features;
+	}
+
+	/** get alpha vector
+	 *
+	 * @return vector to compute posterior mean of Gaussian Process:
+	 *
+	 * \f[
+	 * \mu = K\alpha
+	 * \f]
+	 *
+	 * where \f$\mu\f$ is the mean and \f$K\f$ is the prior covariance matrix.
+	 */
+	virtual SGVector<float64_t> get_alpha();
+
+	/** get Cholesky decomposition matrix
+	 *
+	 * @return Cholesky decomposition of matrix:
+	 *
+	 * \f[
+	 * L = Cholesky(sW*K*sW+I)
+	 * \f]
+	 *
+	 * where \f$K\f$ is the prior covariance matrix, \f$sW\f$ is the vector
+	 * returned by get_diagonal_vector(), and \f$I\f$ is the identity matrix.
+	 */
+	virtual SGMatrix<float64_t> get_cholesky();
+
+	/** update all matrices */
+	virtual void update()=0;
+
+	/** set the noise for inducing points
+	 *
+	 * @param noise noise for inducing points
+	 *
+	 * The noise is used to enfore the kernel matrix about the inducing points are positive definite
+	 */
+	virtual void set_inducing_noise(float64_t noise);
+
+	/** get the noise for inducing points
+	 *
+	 * @return noise noise for inducing points
+	 *
+	 */
+	virtual float64_t get_inducing_noise();
+
+	/** returns derivative of negative log marginal likelihood wrt latent features (input)
+	 * Note that in order to call this method, kernel must support FITC inference
```

Comment:
it is done in the base class for single fitc inference (regression and binary classification) since in the future this class may support multi-kernels.